pyanaconda/ui/gui/spokes/subscription.py

```diff
@@ -0,0 +1,750 @@
+
+        self._authentication_method = AuthenticationMethod.USERNAME_PASSWORD
+
+        self._registration_error = ""
+        self._registration_phase = None
+        self._registration_controls_enabled = True
+
+    # common spoke properties
+
+    @property
+    def ready(self):
+        """The subscription spoke is always ready."""
+        return True
+
+    @property
+    def status(self):
+        # TODO: status message depends on registration/unregistration handling
+        # - shows registration phases when registration + subscription is ongoing
+        # - otherwise shows not-registered/registered/error
+        return ""
+
+    @property
+    def mandatory(self):
+        """The subscription spoke is mandatory if Red Hat CDN is set as installation source."""
+        # TODO: depends on source switching, mandatory if Red Hat CDN is set as installation source
+        return True
+
+    @property
+    def completed(self):
+        return self.subscription_attached
+
+    @property
+    def sensitive(self):
+        # the Subscription spoke should be always accessible
+        return True
+
+    # common spoke methods
+
+    def apply(self):
+        log.debug("Subscription GUI: apply() running")
+        # Set System Purpose data to the Subscription DBus module,
+        # it makes sense to send the rest only right before a
+        # registration attempt.
+        self._set_system_purpose_data_to_subscription_module()
+
+    def refresh(self):
+        log.debug("Subscription GUI: refresh() running")
+        # update subscription request from the Subscription module
+        self._subscription_request = self._get_subscription_request()
+
+    # general purpose properties
+
+    @property
+    def subscription_request(self):
+        """A mirror of the subscription request from the Subscription DBus module.
+
+        Should be always set and is periodically updated on refresh().
+
+        :return: up to date subscription request
+        :rtype: SubscriptionRequest instance
+        """
+        return self._subscription_request
+
+    # properties mirroring entries
+
+    @property
+    def custom_server_hostname(self):
+        return self._custom_server_hostname_entry.get_text()
+
+    @custom_server_hostname.setter
+    def custom_server_hostname(self, hostname):
+        self._custom_server_hostname_entry.set_text(hostname)
+
+    @property
+    def custom_server_hostname_set(self):
+        # even if custom server hostname is set, this can be overridden
+        # by the server checkbox
+        return self._custom_server_hostname_checkbox.get_active()
+
+    @property
+    def http_proxy_location(self):
+        return self._http_proxy_location_entry.get_text()
+
+    @http_proxy_location.setter
+    def http_proxy_location(self, location):
+        self._http_proxy_location_entry.set_text(location)
+
+    @property
+    def http_proxy_username(self):
+        return self._http_proxy_username_entry.get_text()
```

Comment:
I am not a big fan of this approach. Instead of keeping the data in UI elements, I would prefer to keep the data in the subscription request. These properties could be replaced with signal handlers that would update the request with a new value of the changed UI element. Then you can always use the request to check the current values.

This works pretty well in the custom partitioning spoke (see https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/ui/gui/spokes/custom_storage.py#L1563). I use the request to log the current state or to check if something has changed.

However, it might be a too big change that could complicate next backports, so it is up to you.

Reply:
I think it is a good idea - this always seemed cumbersome, but with the DBus module infrastructure on RHEL 8.2 & the available timeframe I was not able to come up with something better.

I like the idea & if its known to work elsewhere I think I can make it work here as well.